ttnn/cpp/ttnn/operations/core/to_layout/to_layout_op.cpp

```diff
@@ -42,6 +42,90 @@ inline bool use_multicore_device_tilize(
     return num_tiles_in_row <= max_tiles;
 }
 
+bool requires_padding_change(const ttnn::Tensor& tensor, ttnn::Layout layout) {
+    auto tile = tensor.get_tensor_spec().tile();
+    if (layout == Layout::ROW_MAJOR) {
+        // There shouldn't be extra paddings for Row Major layout
+        return tensor.logical_shape() != tensor.padded_shape();
+    }
+    // It's okay for conversion to tile layout to preserve arbitrary padding as long as it satisfies the alignment
+    TensorSpec padded_spec(
+        tensor.padded_shape(),
+        TensorLayout(tensor.dtype(), PageConfig(layout, std::move(tile)), tensor.memory_config()));
+    return tensor.get_padded_shape() != padded_spec.padded_shape();
+}
+
+template <typename T>
+Tensor to_layout_impl_on_device(
+    const ttnn::Tensor& tensor_arg,
+    const ttnn::Layout layout,
+    const std::optional<ttnn::DataType>& dtype,
+    ttnn::MemoryConfig output_memory_config,
+    T* device) {
+    bool use_multicore_untilize = true;
+    bool use_multicore_tilize = use_multicore_device_tilize(tensor_arg, dtype);
+
+    if (layout == ttnn::ROW_MAJOR_LAYOUT) {
+        TT_FATAL(
+            !dtype.has_value() || dtype.value() == tensor_arg.dtype(),
+            "dtype cannot be different from tensor dtype when converting to ROW_MAJOR_LAYOUT on device!");
+    }
+
+    if (!requires_padding_change(tensor_arg, layout)) {
+        if (layout == ttnn::ROW_MAJOR_LAYOUT) {
+            return ttnn::untilize(tensor_arg, output_memory_config, use_multicore_untilize);
```

Comment:
ok to call even if tensor was in tile layout?

Reply:
I think untilize is supposed to be called for a tensor in tile layout